src/Semantic/Git.hs

```diff
@@ -17,62 +17,74 @@ module Semantic.Git
 
 import Prologue
 
-import           Data.Attoparsec.Text (Parser)
-import           Data.Attoparsec.Text as AP
+import           Data.Attoparsec.ByteString (Parser)
+import           Data.Attoparsec.ByteString as AP
+import           Data.ByteString (ByteString)
+import           Data.ByteString.Internal (w2c)
+import qualified Data.ByteString.UTF8 as UTF8
+import qualified Data.ByteString as B
 import qualified Data.ByteString.Streaming as ByteStream
+import qualified Data.Attoparsec.ByteString.Streaming as AP.Streaming
 import           Data.Char
 import           Data.Either (fromRight)
 import           Data.Text as Text
-import           Shelly hiding (FilePath)
+import           Text.Parser.Combinators (sepEndBy)
 import qualified Streaming.Process
-import qualified System.Process as Process (proc)
+import qualified System.Process as Process
 import qualified Source.Source as Source
 
 -- | git clone --bare
 clone :: Text -> FilePath -> IO ()
-clone url path = sh $ do
-  run_ "git" ["clone", "--bare", url, pack path]
+clone url path = Process.callProcess "git"
+  ["clone", "--bare", Text.unpack url, path]
 
 -- | Runs @git cat-file -p@. Throws 'ProcessExitedUnsuccessfully' if the
 -- underlying git command returns a nonzero exit code. Loads the contents
 -- of the file into memory all at once and strictly.
 catFile :: FilePath -> OID -> IO Source.Source
 catFile gitDir (OID oid) =
-  let process = Process.proc "git" ["-C", gitDir, "cat-file", "-p", Text.unpack oid]
+  let process = Process.proc "git" ["-C", gitDir, "cat-file", "-p", UTF8.toString oid]
       consumeStdout stream = Streaming.Process.withProcessOutput stream ByteStream.toStrict_
   in Source.fromUTF8 <$> Streaming.Process.withStreamProcess process consumeStdout
 
 -- | git ls-tree -rz
 lsTree :: FilePath -> OID -> IO [TreeEntry]
-lsTree gitDir (OID sha) = sh $ parseEntries <$> run "git" ["-C", pack gitDir, "ls-tree", "-rz", sha]
+lsTree gitDir (OID sha) =
+  let process        = Process.proc "git" ["-C", gitDir, "ls-tree", "-rz", UTF8.toString sha]
+      allEntries     = (entryParser `sepEndBy` AP.word8 0) <* AP.endOfInput
+      ignoreFailures = fmap (fromRight [] . fst)
+  in Streaming.Process.withStreamProcess process $
+     \stream -> Streaming.Process.withProcessOutput stream (ignoreFailures . AP.Streaming.parse allEntries)
```

Comment:
Parsing in constant space! Should provide nice memory savings for large repos. (Kubernetes’s output for this invocation is over a million characters).